`A kód futtatható verziójának a megnyitása:`

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PythonVilag/Erettsegi_feladatok/blob/main/2019/2019_I_Ceges_autok/Megoldas/cegesauto.ipynb)

## Feladat: [2019_I_Ceges_autok](https://github.com/PythonVilag/Erettsegi_feladatok/blob/main/2019/2019_I_Ceges_autok/Erettsegi/Ceges_autok_FL.pdf) <br> <br>


Készítette: Mizsák Dániel, Pecsenye Samu

Minden jellegű visszajelzést örömmel fogadok az [alábbi e-mail címen.](mailto:info@pythonvilag.hu)

# Szükséges adatfájlok letöltése a futtató környezetbe

In [ ]:
!wget https://raw.githubusercontent.com/PythonVilag/Erettsegi_feladatok/main/2019/2019_I_Ceges_autok/Forrasok/autok.txt

--2022-09-20 19:32:28--  https://raw.githubusercontent.com/PythonVilag/Erettsegi_feladatok/main/2019/2019_I_Ceges_autok/Forrasok/autok.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8085 (7.9K) [text/plain]
Saving to: ‘autok.txt’

autok.txt           100%[===================>]   7.90K  --.-KB/s    in 0s      

2022-09-20 19:32:28 (60.7 MB/s) - ‘autok.txt’ saved [8085/8085]



# 1. Feladat

A bementi szövegfájlból egy list-of-dicts struktúrát hozunk létre. Ez egy kicsit több gépelést igényel mint egy list-of-lists, de mivel a változókra névvel, nem pedig indexszel hivatkozunk, sokkal átláthatóbb.

A sort `str.split()`-tel szétbontjuk, elmentjük egy dictionary-be. Vegyük észre, hogy a *nap*, *ID*, és *be* értékeket nem vátjuk át int-re, mivel azokkal késöbb nem számolunk, viszont a *km*-t igen, mivel azokat késöbb ki fogjuk vonni egymásból.

In [ ]:
print('1. Feladat')
with open('autok.txt', 'r') as bemeneti_file:
    bemenet = list()
    for sor in bemeneti_file:
        nap, oraperc, rendszam, ID, km, be = sor.split()
        uj_sor = dict()
        uj_sor['nap'] = nap
        uj_sor['oraperc'] = oraperc
        uj_sor['rendszam'] = rendszam
        uj_sor['ID'] = ID
        uj_sor['km'] = int(km)
        uj_sor['be'] = be
        bemenet.append(uj_sor)

1. Feladat


# 2. Feladat

Mivel a feladat szövege alapján a bejegyzések időrenben vannak, az utolsó - elvitelhez tartozó- bejegyzés az utolsó idő szerint is.

Nem felejtjük el kezelni azt az esetet sem, ha egy autót sem vittek ki (a bemeneti file teljesen üres).

In [ ]:
print('2. feladat')
f2_nap = None
f2_rendszam = None
for sor in bemenet:
    if sor['be'] == '0': f2_nap, f2_rendszam = sor['nap'], sor['rendszam']

if f2_nap is None:
    print('Egyik autót sem vitték ki a hónapban')
else:
    print(f'{f2_nap}. nap, rendszam: {f2_rendszam}')

2. feladat
30. nap, rendszam: CEG300


Alternatív megoldás: egyszerűbb, ha visszafelé megyünk végig a bejegyzések listáján, amíg egy kivitelhez tartozó bejegyzést nem találunk. Ezután akár abba is hagyhatjuk a keresést.

```python
print('2. feladat alternativ')
for sor in reversed(bemenet):
    if sor['be'] == '0':
        print(f"{sor['nap']}. nap, rendszam: {sor['rendszam']}")
        break
else:
    print('Egyik autot sem vittek ki')
```

A printnél a külső és belső idézőjelek különbözőek kell, hogy legyenek, hogy a fordító/interpreter tudja, hogy nem az f-stringet akarjuk lezárni, hanem azon belül akarunk kezdeni egy beágyazott stringet

# 3. Feladat

Ugyan úgy, mint a 2. feladatban, a bekért nap számát nem váltjuk át számmá feleslegesen.

Mivel a mintában az értékek nincsenek szövegbe foglalva, egyszerűbb a több argumentumos print()-et használni, az f-stringek helyett.

In [ ]:
print('3. feladat')
f3_nap = input('Nap [1-30]: ')
print(f'Forgalom a(z) {f3_nap}. napon:')
for sor in bemenet:
    if sor['nap'] == f3_nap:
        be_ki = 'be' if sor['be'] == '1' else 'ki'
        print(sor['oraperc'], sor['rendszam'], sor['ID'], be_ki)

3. feladat
Nap [1-30]: 4
Forgalom a(z) 4. napon:
12:50 CEG303 561 ki
19:17 CEG308 552 be


# 4. Feladat

In [ ]:
print('4. feladat')
f4_szamlalo = 0
for sor in bemenet:
    if sor['be'] == '0':
        f4_szamlalo += 1
    else:
        f4_szamlalo -= 1
print(f'A honap vegen {f4_szamlalo} autot nem hoztak vissza.')

4. feladat
A honap vegen 4 autot nem hoztak vissza.


# Struktúraváltás

Az 5-7. feladatok egyszerűbb megoldásához a list-of-dicts bemeneti strukturát átalakítjuk egy dict-of-list-of-dicts-re.

Minden rendszámhoz tartozni fog egy, a régihez hasonló lista, amiben csak az adott rendszámhoz tartozó bejegyzések szerepelnek, tartva a relatív sorrendet.

In [ ]:
bemenet_valogatott = dict()
for sor in bemenet:
    if sor['rendszam'] not in bemenet_valogatott:
        bemenet_valogatott[sor['rendszam']] = list()
    bemenet_valogatott[sor['rendszam']].append(sor)

# 5. Feladat

Egy autó által megtett utat az első és utolsó km-óra-állás alapján számolhatjuk ki. A -1 index egy indexelt tömbben mindig az utolsó elemre mutat (ha az nem üres), -2 az utolsó előttire, stb.

In [ ]:
print('5. feladat')
for rendszam, sorok in bemenet_valogatott.items():
    f5_elso = sorok[0]['km']
    f5_utolso = sorok[-1]['km']
    f5_megtett = f5_utolso - f5_elso
    print(f'{rendszam} {f5_megtett} km')

5. feladat
CEG306 7165 km
CEG304 6564 km
CEG302 5101 km
CEG308 6745 km
CEG307 6489 km
CEG309 1252 km
CEG303 7465 km
CEG301 5441 km
CEG305 5232 km
CEG300 6751 km


A feladat szövege szerint az nem baj, hogy a rendszámok nem betürendben vannak, hanem az első megjelenésük szerint a bemeneti listában.

De ha mégis rendezni akarnánk, a `sorted()` függvénnyel ez is megoldható.



```python
print('5. feladat rendezett')
for rendszam, sorok in sorted(bemenet_valogatott.items()):
    f5_elso = sorok[0]['km']
    f5_utolso = sorok[-1]['km']
    f5_megtett = f5_utolso - f5_elso
    print(f'{rendszam} {f5_megtett} km')
```


# 6. Feladat

Az egy kivitel alatt megtett utat úgy számoljuk, hogy kivitelkol megjegyezzük a km-óra állását, visszahozatalkol pedig kivonjuk az akkori állásból.

Másik megoldás az lehetne, hogy nem soronként megyünk végig az adatunkon egy if else elágazást használva, hanem két sort együtt kezelve (mivel az autónként válogatott listában garantált, hogy a ki- és bemenetek felváltva szerepelnek). Ez viszont bonyolultabb, és több lehetséges hibával jár.

A választott módszer 'magától' kezeli azt az esetet is, ha egy autót a hónap végén nem hoztak vissza.

In [ ]:
print('6. feladat')
f6_max_km = -1
f6_ID = None
for rendszam, sorok in bemenet_valogatott.items():
    for sor in sorok:
        if sor['be'] == '0':
            f6_km_ki = sor['km']
        else:
            f6_km_be = sor['km']
            f6_megtett = f6_km_be - f6_km_ki
            if f6_megtett > f6_max_km:
                f6_max_km = f6_megtett
                f6_ID = sor['ID']
print(f'Leghosszabb út: {f6_max_km} km, személy: {f6_ID}')

6. feladat
Leghosszabb út: 1551 km, személy: 506


# 7. Feladat




Mivel az adatsort autónként már szétválogattuk, az egyetlen trükköt a formázás jelenti. Mint a 6. feladatban, itt is kezelhetnénk az adatot kétsoronként, de ez a megoldás kompaktabb. A kivitelekhez tartozó sorok után nem rakunk új sor karaktert ('\n'), így a következő iteráció ugyanahhoz a sorhoz fog hozzáírni.

Figyeljünk arra, hogy a példához hasonló sorrendben, formázással írjuk ki az adatokat.

Itt is ellenőrizzük, hogy a kimeneti fájlok megvannak (mindegyik), jó névvel hoztuk őket létre, valamint a tartalmuk is megfelel a példának.

Ez a megoldás csak azokhoz az autókhoz készít menetlevelet, amelyeket legalább egyszer elvittek a hónap folyamán. A többi autóhoz készíthetnénk üres fájlokat, viszont mivel a feladat a flotta rendszámait nem adja meg, csak a példában szerepelnek, ezért nem tudunk nekik menetlevelet írni.

In [ ]:
for rendszam, sorok in bemenet_valogatott.items():
    with open(f'{rendszam}_menetlevel.txt', 'w') as kimeneti_file:
        for sor in sorok:
            if sor['be'] == '0':
                print(f"{sor['ID']}\t{sor['nap']}. {sor['oraperc']}\t{sor['km']} km\t",
                      end='', file=kimeneti_file)
            else:
                print(f"{sor['nap']}. {sor['oraperc']}\t{sor['km']} km\t",
                      end='\n', file=kimeneti_file)

Az f-stringek nem csak `print()` függvényen belül használhatóak, bármilyen kontextusban tudunk velük szöveget formázni, akár az open() függvényen belül is.

A `print()` három kwarg (keyword argument)-je:
*   sep: a printen belüli különböző stringek közötti elválasztó karakter, alapból egy darab szóköz ('\s')
*   end: a print vége után rakott elválasztó karakter, alapból egy új sor karakter ('\n')
*   file: a print célfájlja, alapból a jelenlegi konzol-ra ír.



In [ ]:
## hardcoded text file name
with open('CEG304_menetlevel.txt', 'r') as file:
  for sor in file:
    print(sor)

583	1. 09:04	8477 km	2. 22:35	9235 km	

516	3. 14:08	9235 km	3. 16:04	9414 km	

551	3. 17:53	9414 km	3. 22:48	9569 km	

546	5. 08:20	9569 km	5. 14:22	9652 km	

517	5. 15:04	9652 km	9. 19:34	11086 km	

551	10. 12:49	11086 km	11. 21:20	11999 km	

574	13. 10:12	11999 km	13. 20:17	12153 km	

549	15. 14:46	12153 km	16. 21:29	12854 km	

591	18. 08:43	12854 km	18. 18:54	13007 km	

534	19. 07:53	13007 km	19. 22:43	13279 km	

560	20. 08:35	13279 km	20. 11:27	13452 km	

588	21. 16:58	13452 km	23. 20:28	14335 km	

512	24. 16:58	14335 km	26. 22:21	15041 km	

504	27. 13:47	15041 km	


Az oszlopok elcsúsznak, ha valamelyik sor nem fér bele 7 karakterbe ('11086 km' 8 karakter hosszú). Ez nem megy szembe a feladat követelményeivel, teljesen jó így.

# Utószó

Amennyiben a megoldás során használt kód megértése problémát okoz, látogasd meg a [Python alapjai kurzusomat](https://pythonvilag.hu/pv/python/alapok/), ahol részletesen bemutatom a Python programozási nyelv alapvető eszköztárát. <br> <br>
További kidolgozott érettségi feladatokat pedig [az alábbi linken](https://pythonvilag.hu/pv/python/alapok/) éred el.
<br> <br>

Találkozunk a következő leckében!